docs/dependency-injection/dependency-injection.md

```diff
@@ -0,0 +1,220 @@
+
+### One container by default 
+
+A typical Oryx application usually uses one global DI container, which is set up by the app orchestrator. This is currently the recommended approach for Oryx applications that utilize web components.
+
+# Defining services
+
+In Oryx, we follow certain conventions for defining services.
+
+## String-Based Token
+
+Services are resolved using string tokens. For example:
+
+```ts
+export const CartService = 'oryx.CartService';
+```
+
+It is recommended to use a project-specific prefix for those tokens to minimize the risk of naming collisions with other services or libraries.
+
+## Service Interface
+
+The service interface is a plain TypeScript interface that describes the public API of the service. Here's an example:
+
+```ts
+export interface CartService {   
+  load(): Observable<null>;
+  getCart(data?: CartQualifier): Observable<Cart | null>; 
+}
+```
+
+Since TypeScript interfaces are lightweight, there's no overhead or tight coupling between the service definition (and its token) and its implementation.
+
+## Coupling Token with Interface
+
+To achieve type safety and couple the string token with a specific interface, we use a technique to augment the global `InjectionTokensContractMap`:
+
+```ts
+declare global {
+  interface InjectionTokensContractMap {
+    [CartService]: CartService;   
+  } 
+}
+```
+
+Thanks to this, both `resolve` and `inject` are able to infer the proper type when injecting a service using a string token, providing type-safety for your code.
+
+## Default Implementation
+
+The default implementation is usually a class that implements the service interface. Here's an example:
+
+```ts
+export class DefaultCartService implements CartService { 
+// implementation 
+}
+```
+
+# Providing services
+
+To use and inject services, they must be provided in the DI container. Oryx offers several types of providers:
+
+-   `ClassProvider<T>`: Binds a token to a class constructor.
+-   `ValueProvider<T>`: Binds a token to a value.
+-   `FactoryProvider<T>`: Binds a token to a factory function.
+
+You can pass a list of providers as an argument to the `Injector` constructor or to the Injector's `provide` method. However, it's important to note that providing services is only possible before the first usage of the DI container.
+
+In Oryx applications, providers are typically grouped into const arrays and passed to the app orchestrator either directly or as part of feature definitions. By doing this, you can easily manage features with corresponding providers and their dependencies.
+
+Here's an example describing how to use the app orchestrator to provide services as part of a feature:
+
+```ts
+export const app = appBuilder()
+  .withFeature(someFeature)
```

Comment:
Maybe let's drop it, to focus on the example.

Reply:
hmm... ok.